### ISSUE TITLE
Support for animated images

### ISSUE THREAD

**Issue by akiross** [slint-ui/slint#2081]
Hello,
in [this comment](https://github.com/slint-ui/slint/issues/1326#issuecomment-1148863995) it was mentioned that a image decoding rework was in progress and that support for loading animated GIFs, webp and apng could be possible.

I would like to know if there are news/plans on that: I recently tested version 0.3.3, trying to load an animated GIF, but I couldn't (invalid file format) and I couldn't find any flag to enable those.

Thanks!

**Comment by tronical** [comment]
Hi @akiross . The work of doing the image decoding in the core library was done, but support for animated GIFs/webp/apng is indeed not implemented yet.

AFAIK this isn't on anyone's short-term todo list, but I think that it could be done in a platform independent way. The `ImageInner` would have to be extended with a new type of image. I'm not sure yet how to store the `image::AnimatedImageDecoder` impl alongside the `Frames<'a>` it can create, but there must be a way.

There's also a way of working around this on the application side, by using `image-rs` directly there and using the `slint::Image` API to create an image from from a buffer - like in the image filter example.

**Comment by ogoffart** [comment]
While we are still working on having such support builtin, it could be done in native code with something like this:

```
global ImageLoader := {
   callback load_gif_frame(string, int) -> image;
}
//...
   Image {
         source: ImageLoader.load_gif_frame("foo.gif", animation-tick() / 1ms);
   }
```

and set a callback that return the right frame in an image.
(some caching would be in order to avoid re-opening the file on each frame)

**Comment by akiross** [comment]
Nice, thank you! I'll try to use the workaround for now.

I don't think I'll have time for this in the short term, but I could try to give a look and maybe contribute to this.

I leave this open since I see it's a possible enhancement.

**Comment by dongjian** [comment]
> While we are still working on having such support builtin, it could be done in native code with something like this:
> 
> ```
> global ImageLoader := {
>    callback load_gif_frame(string, int) -> image;
> }
> //...
>    Image {
>          source: ImageLoader.load_gif_frame("foo.gif", animation-tick() / 1ms);
>    }
> ```
> 
> and set a callback that return the right frame in an image. (some caching would be in order to avoid re-opening the file on each frame)

Hello, I tried to write this but got an error message  ``` Call of impure callback 'load-gif-frame' ```   how to avoid this error? Thank you for your help

**Comment by tronical** [comment]
I'm starting to think that perhaps we should indeed support webp/apng/gif out of the box (using image-rs) and per opt-in feature also video formats supported by ffmpeg. So if you enable the `slint/ffmpeg` feature we could also support those, without sound.

API wise, an option would be to have a `AnimatedImage` element, because you might want to have an additional boolean `running` property to start/stop the playback.

**Comment by ogoffart** [comment]
API wise, an option would be to have a AnimatedImage element

We can just add `running` to the Image element. I don't think we need another element for that.

**Comment by MiKom** [comment]
> I'm starting to think that perhaps we should indeed support webp/apng/gif out of the box (using image-rs) and per opt-in feature also video formats supported by ffmpeg. So if you enable the `slint/ffmpeg` feature we could also support those, without sound.

I agree. Having a multimedia system would definitely be much heavier. For webp/apng/gif you probably don't even need to introduce new dependencies because image-rs should already support what's needed.

Having a separate `AnimatedImage` component may make sense because there would be a bit of API there. Starting/stopping is one thing, but also controlling the looping behavior, playback speed, seeking etc.
This could potentially unnecessary pollute the simple `Image` type. Probably not everything at the start, but it may grow.

**Comment by receiver1** [comment]
> While we are still working on having such support builtin, it could be done in native code with something like this:
> 
> ```
> global ImageLoader := {
>    callback load_gif_frame(string, int) -> image;
> }
> //...
>    Image {
>          source: ImageLoader.load_gif_frame("foo.gif", animation-tick() / 1ms);
>    }
> ```
> 
> and set a callback that return the right frame in an image. (some caching would be in order to avoid re-opening the file on each frame)

How I can do this in C++?
